### PR TITLE
First draft of the conformance guide

### PR DESCRIPTION
**Pull request by phaumer**
Signed-off-by: Peter Haumer <4391934+phaumer@users.noreply.github.com>

## Release Notes

Milestone:

Changelog:

## Types of changes

What types of changes does your code introduce to Zowe Explorer?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] Updates to Documentation or Tests (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This checklist will be used as reference for both the contributor and the reviewer_

- [ ] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/vscode-extension-for-zowe/wiki/Best-Practices:-Contributor-Guidance) wiki
- [ ] PR title follows [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0-beta.2/)
- [ ] PR Description is included
- [ ] gif or screenshot is included if visual changes are made
- [ ] `yarn workspace vscode-extension-for-zowe vscode:prepublish` has been executed
- [ ] All checks have passed (DCO, Jenkins and Code Coverage)
- [ ] I have added unit test and it is passing
- [ ] I have added integration test and it is passing
- [ ] There is coverage for the code that I have added
- [ ] I have tested it manually and there are no regressions found
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any PR dependencies have been merged and published (if appropriate)

## Further comments

As discussed in meeting 4/27 we can use PRs to edit this doc as a team now.

### CODE REVIEW

**Review by dkelosky — COMMENTED** [review]
It looks like we have two proposals for v1 conformance that have different requirements.  The other is here: https://ent.box.com/s/hqzfsia30z09hsjo8gj40kunhjaw9ebj .  Which is active?  

It would be nice to have conformance criteria sourced alongside the repo; however, this format is different than conformance for other components.  Will Zowe Explorer conformance look too different than other parts of Zowe?

**Comment by phaumer** [comment]
@dkelosky I converted the Excel sheet into this doc as it can be more verbose and it better for discussion as I observed not everyone fully understood each criteria in the same way. Pls the sheets are unreadable in Box making it the worst collaboration tool to use for this. In the PR we can provide line by line feedback now. Once we agreed on the content we will paste back to the sheet if Zowe requires it.

Do you see other areas that are different? One thing we wanted to add was to define how each criteria can be verified at least by the submitter, but also be the approval board, if necessary.

**Comment by dkelosky** [comment]
The idea of an approval board itself is also a different process than other parts of Zowe.  @RASakach might have some comments on this piece.  

A few things I noticed in here versus the excel doc:
- the high level areas of `a`, `b`, and `c` in the excel doc are named differently than `1`, `2`, and `3` - are they intended to be the same? 
- the `General` section doesn't appear to be in the excel doc
- this looks to be based on v2 standards, e.g. CLI does not need to be installed is required for v2, but best practice for v1 in excel, but this PR shows it as required

Sorry to be a pain; I hope I'm looking at the right things 😄

**Comment by RASakach** [comment]
Hi Dan - thanks for looping me in here.  The Conformance Program is managed at the OMP level, not at the Zowe project level.  So the squads are not authorized to approve or disapprove conformance applications.  That said, if the ZE Squad would like to provide the OMP with a means to verify an applicant's criteria - that is something we can consider adding BUT, OMP must agree to taking on this burden.

Also - to be clear - the format for all Zowe Conformance Programs is a "table" which can easily been rendered into a checklist.  So, I would recommend staying in that format.  It is reasonable that many entries may require more details for the applicant.  In this case we recommend posting a separate document with Zowe Explorer Conformance Criteria Details and using a LINK to that document for those criteria entries that need to reference it.  Finally, we realize a BOX drive is not ideal for collaboration -- if this becomes a hardship we can work towards finding alternative means storing these WIP documents provided all community members have the ability to use it (e.g. I believe some vendors do not have the ability to use Google products).  

Hope this helps.

**Comment by phaumer** [comment]
@dkelosky yes, my proposal was to cover the same low-level items, but to structure them differently as the current structure causes confusion and the wrong things (at least how I interpreted them) were listed in the wrong groups as well. As I am not good at Excel sheet refactoring I was trying this, also for the other reasons mentioned. I also proposed different required vs best practices assignment and explained the reason in our Tu scrum.

The next steps we discussed as a team yesterday and today was to write one sentence summaries for each criteria and paste them back into the sheet (the second one with the 2 in the name?) I will then also write a rationale summary for why certain changes were made.

**Comment by phaumer** [comment]
@RASakach I would propose a github repo and MD documents, which also can render tables; but often bullet point lists are more readable. The advantage is that people can create Pull Requests and reviewers can comment line by line with the excellent Github tools.

**Comment by dkelosky** [comment]
Thanks @phaumer - I do see these as somewhat similar but ultimately, two different proposals since the format and values changed.  Is the intent of this markdown proposal to disregard the excel version?  

I think we need to have a similar check list to the other Zowe components.  I like your idea of markdown, but perhaps this is something that should be pushed out to all squads simultaneously in a future PI?

**Comment by phaumer** [comment]
Once we have agreement on the contents of this file then I will paste back to the sheet.

**Comment by dkelosky** [comment]
> Once we have agreement on the contents of this file then I will paste back to the sheet. 

Thanks - so the intent is that this supersedes the excel document.  We've had other Zowe VS Code extension developers internally and external to the squad review the excel document.  I want to re-share this new draft if they need to review again.

**Comment by dkelosky** [comment]
table format to reference:
https://github.com/openmainframeproject/foundation/blob/main/zowe_conformance/test_evaluation_guide_table.md#zowe-cli---zowe-v1

**Comment by phaumer** [comment]
@jellypuno @dkelosky as discussed I extracted one sentence that could be used for the conformance text. I also created a new version of the Excel sheet in the Box folder and pasted the content there: <https://ent.box.com/s/id5k7hcpfiyried8xape1o75w44obfyi>

**Comment by jellypuno** [comment]
Thank you @phaumer. I am adding my comments under proposed changes. I need to rephrase the criteria in a form that the onboarding squad recommended.

**Comment by dkelosky** [comment]
thanks @phaumer

File: docs/Conformance-Guide-draft.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+## General conformance criteria (0)
+
+These criteria are independent of the kind of extension produce and need to be fulfill for all cases.
+
+1. (Required) Naming: **If the extension uses the word "Zowe" in its name, it abides by Linux Foundation's Trademark Usage rules to ensure the word Zowe is used in a way intended by the Zowe community.**
+
+   The extender needs to provide a link where the offering will be announced and/or made available to the public.
+
+1. (Required): No Zowe CLI plugin installation requirement: **If the extender makes use of a Zowe CLI profile other than the Zowe Explorer default `zosmf` then the extension must not make any assumptions that a matching Zowe CLI plugin has been installed in the Zowe Explorer user's environment.**
```
Comment (jellypuno) [inline docs/Conformance-Guide-draft.md:31]: **No Zowe CLI plugin installation dependency:** If the extension makes use of a Zowe CLI profile other than the Zowe Explorer default `zosmf`, it must not make any assumptions that a matching Zowe CLI plugin has been installed in the Zowe Explorer user's environment.

File: docs/Conformance-Guide-draft.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+   In other words the extension must be full self-contained including all the code of the Zowe CLI Plugin that implements the new profile. This will not only simplify the end user experience, but also ensures that the extension can run in other VS Code compatible environments such as Cloud IDEs such as Eclipse Che. For VS Code extensions Zowe Explorer provides APIs to call that ensure that the users can store and access such new profiles in Zowe home directory folders and secure credentials store, which should be used or an equivalent needs to be provided. To test this requirement a user shall be able start the extension and use it without having Nodejs and Zowe CLI installed locally.
+
+1. (Required) Publication tag: **If the extension is published in a public catalog or marketplace such as Npmjs, Open-VSX, or VS Code Marketplace then it must use the tag or keyword "Zowe" so it can be found when searching for Zowe and be listed with other Zowe offerings.**
```
Comment (jellypuno) [inline docs/Conformance-Guide-draft.md:35]: **Publication tag:** If the extension is published in a public catalog or marketplace such as Npmjs, Open-VSX, or VS Code Marketplace, it uses the tag or keyword "Zowe" so it can be found when searching for Zowe and be listed with other Zowe offerings.

File: docs/Conformance-Guide-draft.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+
+   For example, if the extension is a VS Code extension then the "keywords" array of the package.json must include the entry "Zowe". This can be verified by looking at the page's Tags section in the VS Code Marketplace or if not yet present by extracting the vsix file and inspecting the package.json file.
+
+1. (Required) Support: **The extension needs to document how a user would get support or file issues for the extension for problems that are not related to Zowe and Zowe Explorer.**
```
Comment (jellypuno) [inline docs/Conformance-Guide-draft.md:39]: Extension has a clear documentation on how a user will be provided support. It has a clear instruction on how to report problems that are not related to the extension.

File: docs/Conformance-Guide-draft.md [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+   The extender needs to document where the user can find this information in the conformance application.
+
+1. (Best Practice) User settings consistency: **For a consistent user experience we recommend that user settings and configuration settings follow the naming conventions as documented in the Zowe Explorer extensibility documentation.**
```
Comment (jellypuno) [inline docs/Conformance-Guide-draft.md:43]: Extension follows the recommended naming convention for the user settings and configuration settings indicated in the Zowe Explorer extensibility documentation

File: docs/Conformance-Guide-draft.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+   The list and documentation of the configuration settings should be made available as part of the compliance application. If the extension is a VS Code extensions then including the package.json file will be sufficient unless other means of configuration settings (such as environment variables) have been used.
+
+1. (Best Practice) Error message consistency: **For a consistent user experience error and log messages should follow the format documented in the Zowe Explorer extensibility documentation.**
```
Comment (jellypuno) [inline docs/Conformance-Guide-draft.md:47]: Extension follows the recommended error message format indicated in the Zowe Explorer extensibility documentation

File: docs/Conformance-Guide-draft.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+A Profiles Access extension is a Zowe Explorer extension that uses the Zowe Extensibility API to conveniently access Zowe CLI profiles loaded by Zowe Explorer itself. This allows the extension to consistently access profile instances of specific types, offer them for edit and updates as well as common refresh operations that apply to all extensions, add more profile types it is using itself for its own custom views (for example a CICS extension adding a CICS explorer view) and other similar use cases related to Zowe CLI profiles. These extensions do **not** have to be VS Code extension if it just wants to use ProfilesCache implementation of Zowe Explorer as all APIs are provided free of any VS Code dependencies. Such an extension could be used for another non-VS Code tool, a Zowe CLI plugin, a Web Server or another technology. However, to access the profiles cache of the actual running VS Code Zowe Explorer the extender needs to be a VS Code extension that has an extension dependency defined to be able to query the extender APIs. Therefore, some of the criteria that are listed here as required are only required if the extender is a VS Code extension.
+
```
Comment (jellypuno) [inline docs/Conformance-Guide-draft.md:58]: Additional entries:
- (Required) **Sharing of profiles with Zowe CLI:** Extensions shares the created profiles between Zowe CLI and VSCode
- (V1: Best Practice, V2: Required) Extension supports base profiles and tokens
- (V1: Best Practice, V2: Required) Extension supports JSON configuration file
- (V1: Best Practice, V2: Required) If the extension supports Zowe's Secure Credential store, it calls the Zowe Explorer-provided method

Comment (jellypuno) [reply docs/Conformance-Guide-draft.md:58]: Another one before moving to dependencies:

The extension is built on top of at least one Zowe Explorer provided API.

[Select all that apply]
a. Dependency (Extension Accessing Profiles)
b. Dependency (Data Provider Extension)
c. Dependency (Extension Adding Menus)

File: docs/Conformance-Guide-draft.md [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+1. (Required if VS Code) VS Code extension dependency: **Zowe Explorer VS Code extensions can only be activated after Zowe Explorer is fully activated itself. Therefore, to ensure the correct activation order and extender must include an extensionDependencies to Zowe Explorer entry in their package.json file.**
```
Comment (jellypuno) [inline docs/Conformance-Guide-draft.md:59]: Extension defines Zowe Explorer as a VSCode extension dependency (Code Example). It includes an extensionDependencies to Zowe Explorer entry in their package.json file and it is only activated once Zowe Explorer is fully activated.

File: docs/Conformance-Guide-draft.md [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+   The package.json file must be provided to verify this criteria.
+
+   ```json
+   "extensionDependencies": [
+    "Zowe.vscode-extension-for-zowe"
+   ],
+   ```
+
+1. (Required if VS Code) Zowe Extender access: **Zowe Explorer VS Code extensions must access the shared Zowe Explorer profiles cache only via the `ZoweExplorerApi.IApiRegisterClient.getExplorerExtenderApi()` API as documented in the Zowe Explorer extensibility documentation.**
```
Comment (jellypuno) [inline docs/Conformance-Guide-draft.md:69]: Extension accesses the shared Zowe Explorer profiles cache via ZoweExplorerApi.IApiRegisterClient.getExplorerExtenderApi() API as documented in the Zowe Explorer extensibility documentation.

File: docs/Conformance-Guide-draft.md [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+
+   API calls to the Explorer Extender API will be logged by Zowe Explorer and be verified by inspecting Zowe Explorer log files.
+
+1. (Required if VS Code) Added Profile Type initialization: **If the extender depends on a new Zowe CLI profile type other than the Zowe Explorer default `zosmf` then an extender needs to call `ZoweExplorerApi.IApiRegisterClient.getExplorerExtenderApi().initialize(profileTypeName)`**
```
Comment (jellypuno) [inline docs/Conformance-Guide-draft.md:73]: If the extension has a dependecny on a new Zowe CLI profile type other than the Zowe Explorer default `zosmf`, it is calling the ZoweExplorerApi.IApiRegisterClient.getExplorerExtenderApi().initialize(profileTypeName)to ensure that the profile type is supported and managed by the extension without a Zowe CLI plugin installed

**Review by jellypuno — CHANGES_REQUESTED** [review]
I also updated the excel sheet. I am still going through the doc so I might add more comments later

File: docs/Conformance-Guide-draft.md [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+
+   Test run logs can be provided to demonstrate the fulfillment of the requirement.
+
+## Menu extension conformance criteria (3)
+
+1. (Required) Menu Names: **If the extension is adding new commands and context menu entries to the Zowe Explorer tree view nodes, it adheres to the Zowe Explorer-provided contextual string format.**
+
+   **TBD**: Define rules and verification criteria.
+
+   _For team discussion_: We currently only have documented for how to add context menus, but not how the commands that implement the menu can actually access the node on which the click occurred to determine all the information needed (e.g. the data set member name and the profile of the node). We need to implement and provide an API for our tree browsers to properly support this and prevent users for just hacking the tree views themselves. Until we have such an API perhaps this extensibility kind needs to be postponed.
```
Comment (lauren-li) [inline docs/Conformance-Guide-draft.md:103]: Colin's [test mule repo](https://github.com/Colin-Stone/zowe-explorer-test-mule/blob/master/src/extension.ts#L51) demonstrates how it's currently possible for extenders to access the information for the node that was clicked on. Here is a snippet of the information I saw his extender accessing, which includes data set name and profile information:

Screen Shot 2021-05-05 at 8 11 38 AM

Perhaps with an API, we could enforce better control over how much and what kind of node information the extenders can access.

However, as extenders seem to already be able to access the information and hook into our context menus, I was thinking maybe we want to provide some guidelines on how we want extenders to define groupings for their context menu items (i.e. in relation to Zowe Explorer's existing context menu groupings).

File: docs/Conformance-Guide-draft.md (hunk 3 above)
Comment (lauren-li) [reply docs/Conformance-Guide-draft.md:39]: I thought this criteria should be "documentation on how to report problems that **are** related to the extension"? (Assuming the goal of this item is that the extenders' users would know how to file issues against the extender if they encounter problems with it, as opposed to filing issues against Zowe Explorer.)

File: docs/Conformance-Guide-draft.md (hunk 4 above)
Comment (lauren-li) [reply docs/Conformance-Guide-draft.md:43]: If we take @jellypuno's suggested revision, I would propose adding the reason for this criteria ("to provide a consistent user experience") into the paragraph below.

File: docs/Conformance-Guide-draft.md (hunk 5 above)
Comment (lauren-li) [reply docs/Conformance-Guide-draft.md:47]: If we take @jellypuno's suggested revision for this, I would propose adding the reason for this criteria ("to provide a consistent user experience") into the paragraph below.

File: docs/Conformance-Guide-draft.md [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+
+1. (Best Practice) Matching Zowe CLI Plugin: **Provide a Zowe CLI Plugin for the data provider's profile type that implements the core capabilities requires for the new protocol that users can then also use to interact with the protocol outside of the Zowe Explorer extension using Zowe CLI commands.**
+
+   A Zowe CLI Plugin is a regular NPM package that then can be completely imported as a dependency into a VS Code extension that represents a data provider for Zowe Explorer. The zFTP Zowe CLI Plugin and Zowe Explorer extension for zFTP are an example for such a pairing. This requirement can be verified by testing the CLI plugin and seeing it being used in the package.json file of the VS Code extension.
+
+1. (Required) Data provider API implementation: **The Zowe Explorer extensibility API provides interfaces for the required operations to be implemented for MVS, JESS and USS. An extender must either fully implement and register at least one of these three interfaces or alternatively throw JavaScript exceptions that provide meaningful error messages to the end-user in the `Error.message` property of the exception that Zowe Explorer will display in a dialog.**
```
Comment (jellypuno) [inline docs/Conformance-Guide-draft.md:89]: Extension fully implements and registers to at least one of the three Zowe Explorer interfaces or alternatively throw exceptions that provide meaningful error messages to the end-user in the 'Error.message' property that will be displayed in a dialog.

File: docs/Conformance-Guide-draft.md [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+
+1. (Required) New Zowe CLI profile type: **A new data provider requires a new Zowe CLI profile type as it is the foundation for registering the APIs that add the new protocol as documented in the extension guideline. When new APIs get registered via the `ZoweExplorerApi.IApiRegisterClient.registerMvsApi()` call the APIs passed as a parameter needs to use a unique profile type name.**
+
+   Such registration events will be logged by Zowe Explorer and can be used to verify to successful implementation of the requirement in the Zowe Explorer logs.
+
+1. (Best Practice) Matching Zowe CLI Plugin: **Provide a Zowe CLI Plugin for the data provider's profile type that implements the core capabilities requires for the new protocol that users can then also use to interact with the protocol outside of the Zowe Explorer extension using Zowe CLI commands.**
```
Comment (jellypuno) [inline docs/Conformance-Guide-draft.md:85]: If the extension has a Zowe CLI counterpart, it must use the functionalities implemented in the CLI. ??

I wanted to say that if the extension has a CLI, then the functionalities will be shared with VSCode (e.g FTP)

I am not sure if this should just under a data provider or it should be under general. I could have a Zowe CLI Plugin but not a data provider. I could just extend the menu and have a matching plugin and I think sharing functionalities with CLI (in general) should be a requirement

File: docs/Conformance-Guide-draft.md (hunk 7 above)
Comment (lauren-li) [reply docs/Conformance-Guide-draft.md:59]: I believe defining Zowe Explorer in the extender's `extensionDependencies` would ensure that the extender is only activated once Zowe Explorer is activated. So perhaps we could shorten the bolded lines to:

"Extension defines Zowe Explorer as a VSCode extension dependency (Code Example). It includes an extensionDependencies entry for Zowe Explorer in its package.json file."

Then, we can move the explanation of the rationale into the paragraph below: "This requirement ensures the correct activation order, with the extender only activating once Zowe Explorer is fully activated."

File: docs/Conformance-Guide-draft.md [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+
+## Data Provider extension conformance criteria (2)
+
+A data provider Zowe Explorer extension provides an alternative protocol for Zowe Explorer to interact with z/OS. The default protocol Zowe Explorer uses is the z/OSMF REST APIs and data provider adds support for another API. For example, the Zowe Explorer extension for zFTP, which is maintained by the Zowe Explorer squad is an example for a Zowe Explorer data provider extension that uses FTP instead of z/OSMF for all of its USS and MVS interactions. To achieve such an extension it uses a Zowe CLI Plugin for FTP that implemented the core interactions and provided them as an SDK andd
+
+1. (Required) New Zowe CLI profile type: **A new data provider requires a new Zowe CLI profile type as it is the foundation for registering the APIs that add the new protocol as documented in the extension guideline. When new APIs get registered via the `ZoweExplorerApi.IApiRegisterClient.registerMvsApi()` call the APIs passed as a parameter needs to use a unique profile type name.**
```
Comment (jellypuno) [inline docs/Conformance-Guide-draft.md:81]: The extension registers its new API via the ZoweExplorerApi.IApiRegisterClient.registerMvsApi() call as indicated from the Zowe Explorer extensibility documentation

**Review by jellypuno — DISMISSED** [review]

File: docs/Conformance-Guide-draft.md (hunk 9 above)
Comment (lauren-li) [reply docs/Conformance-Guide-draft.md:73]: @jellypuno's suggestion: Perhaps just list the requirement in bold, and then explain the rationale in the paragraph below, e.g.:

>If the extension has a dependency on a new Zowe CLI profile type other than the Zowe Explorer default `zosmf`, then it calls `ZoweExplorerApi.IApiRegisterClient.getExplorerExtenderApi().initialize(profileTypeName)`.
>
> This requirement ensures that the profile type is supported and managed by the extension without a Zowe CLI plugin installed...

File: docs/Conformance-Guide-draft.md (hunk 13 above)
Comment (lauren-li) [reply docs/Conformance-Guide-draft.md:81]: Building on @jellypuno's suggestion: Move rationale to the paragraph below, rather than removing it completely.

**Comment by dkelosky** [comment]
fyi @davidkjackson54

**Comment by dkelosky** [comment]
fyi @robertblum-psi

File: docs/Conformance-Guide-draft.md (hunk 13 above)
Comment (phaumer) [reply docs/Conformance-Guide-draft.md:81]: I think we need to point out that the registration of new APIs work with a newly defined profile type, i.e. you cannot register an alternative zosmf implementation, for example.

**Comment by robertblum-psi** [comment]
@dkelosky asked me to review and comment.   @dkelosky: Thanks for the heads up. 

**General conformance criteria (0) Item 2.**
This item _should not_ be Required, but is okay a Best Practice.  It is ridiculous to assume that an integrator is going to take the time to re-invent the wheel when a perfectly functional, potentially very complex, CLI already exists.   Why they should do so as a _best practice_ is obvious, but that doesn't mean users won't be willing to use the plug-in in restricted number of environments or after having to install Nodejs and the Zowe CLI.  Enforcing this will pretty much prevent adoption of promote the standard and the proliferation of non-conformant extensions.

This item seems to conflict with **Data Provider extension conformance criteria (2)**, which seems to imply leveraging another CLI: "zFTP... uses a Zowe CLI Plugin.."

**General conformance criteria (0) Item 7.**
Even though this is in the category of suggestions, I'm a vendor and we have better performing code the bypasses z/OS and spooler code.  Other vendors will, too.  This is very unfriendly.

**Data Provider extension conformance criteria (2) Item 4**
This is fine as a Best Practice.  It should not be promoted to Required as it can impede rapid development

**Comment by robertblum-psi** [comment]
**General conformance criteria (0)**
I highly suggest that a minimal standard of documentation be Required for conformance.  Under no circumstances should having to refer to the code be considered conformant.  At minimum, the documentation should explain what the extension does, how the end user should interact to get the result, and give one or more examples.  This document can be built into the extension, delivered with it in text or PDF that the end user can access easily, or be a link on the builder's website.

File: docs/Conformance-Guide-draft.md (hunk 4 above)
Comment (jellypuno) [reply docs/Conformance-Guide-draft.md:43]: How do I add it and still make the sentence a Y or N question? Like this?

Extension provides a consistent user experience and follows the recommended naming convention for the user settings and configuration settings indicated in the Zowe Explorer extensibility documentation.

**Comment by jellypuno** [comment]
I've updated the table md file. I feel like I wasn't able to add all of @lauren-li comments. If so, please kindly update it. Thank you! 

For General Extension item 7 & 8 - I am still if it should be in general extension or under profiles. @dkelosky what do you think?

I also wanted  to consider @robertblum-psi suggestion regarding the documentation of the extension's architecture but based on our previous discussions I don't think LF  will review it.

**Comment by lauren-li** [comment]
> I've updated the table md file. I feel like I wasn't able to add all of @lauren-li comments. If so, please kindly update it. Thank you!

@jellypuno Do you want to apply the updates you made for the table to the text document, as well? My comments were for the conformance draft text document (not the table), and were largely conditional on your suggested revisions being applied there.

File: docs/Conformance-Guide-draft.md (hunk 3 above)
Comment (phaumer) [reply docs/Conformance-Guide-draft.md]: Extension has documentation with instructions on how to report problems that are related to the extension and not Zowe Explorer. It needs to explain how users can determine if a problem is related to the extension or Zowe Explorer.

File: docs/Conformance-Guide-draft.md [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+
+   For example, if the extension is a VS Code extension then the "keywords" array of the package.json must include the entry "Zowe". This can be verified by looking at the page's Tags section in the VS Code Marketplace or if not yet present by extracting the vsix file and inspecting the package.json file.
+
+1. (Required) Support: **The extension needs to document how a user would get support or file issues for the extension for problems that are not related to Zowe and Zowe Explorer.**
+
+   The extender needs to document where the user can find this information in the conformance application.
+
+1. (Best Practice) User settings consistency: **For a consistent user experience we recommend that user settings and configuration settings follow the naming conventions as documented in the Zowe Explorer extensibility documentation.**
+
+   The list and documentation of the configuration settings should be made available as part of the compliance application. If the extension is a VS Code extensions then including the package.json file will be sufficient unless other means of configuration settings (such as environment variables) have been used.
```
[the remark this answers is not in the document]
Comment (phaumer) [reply docs/Conformance-Guide-draft.md:45]: As it is Best Practice would do that in the guide.

File: docs/Conformance-Guide-draft.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (phaumer) [reply docs/Conformance-Guide-draft.md]: The point is rather the extender providing a new type with a name not used, yet. The planned init method could do that check.

File: docs/Conformance-Guide-draft.md (hunk 12 above)
Comment (phaumer) [reply docs/Conformance-Guide-draft.md]: Made this a best practice as one could write a data provider without CLI.

File: docs/Conformance-Guide-draft.md (hunk 13 above)
Comment (phaumer) [reply docs/Conformance-Guide-draft.md]: Extension registers its new API instances with a new profile type name for the different Zowe Explorer views via the `ZoweExplorerApi.IApiRegisterClient.register{Mvs|Uss|Jes}Api(profileTypeName)` call as indicated from the Zowe Explorer extensibility documentation

**Comment by phaumer** [comment]
> **General conformance criteria (0) Item 7.**
> Even though this is in the category of suggestions, I'm a vendor and we have better performing code the bypasses z/OS and spooler code. Other vendors will, too. This is very unfriendly.

Fair point. I added a qualifying sentence to use the SDK as a default, but not if the intention of the extension is to actually replace the SDK.

**Comment by phaumer** [comment]
> **General conformance criteria (0)**
> I highly suggest that a minimal standard of documentation be Required for conformance. Under no circumstances should having to refer to the code be considered conformant. At minimum, the documentation should explain what the extension does, how the end user should interact to get the result, and give one or more examples. All features and options need be covered. This document can be built into the extension, delivered with it in text or PDF that the end user can access easily, or be a link on the builder's website.

Yes, Zowe Conformance is self-evaluation only. No one else other than the authors of the extension will certify compliance. We will add some helping features such as logging of API calls so extension developers can more easily check that their calls succeeded.

File: packages/zowe-explorer-api/src/tree/ZoweVsCodeExtension.ts [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+     * @returns an initialized instance `ZoweExplorerApi.IApiRegisterClient` that extenders can use
+     *          to access the Zowe Explorer APIs or `undefined`. Also `undefined` if requiredVersion
+     *          is larger than the version of Zowe Explorer found.
+     */
+    public static getZoweExplorerApi(requiredVersion?: string): ZoweExplorerApi.IApiRegisterClient {
+        const zoweExplorerApi = vscode.extensions.getExtension("Zowe.vscode-extension-for-zowe");
+        if (zoweExplorerApi?.exports) {
+            const zoweExplorerVersion =
+                ((zoweExplorerApi.packageJSON as Record<string, unknown>).version as string) || "15.0.0";
+            if (requiredVersion && semver.valid(requiredVersion) && !semver.gte(zoweExplorerVersion, requiredVersion)) {
```
Comment (lauren-li) [inline packages/zowe-explorer-api/src/tree/ZoweVsCodeExtension.ts:32]: @phaumer Should `!semver.gte` be `!semver.get`?

Comment (phaumer) [reply packages/zowe-explorer-api/src/tree/ZoweVsCodeExtension.ts:32]: @lauren-li gte stands for greater-or-equal.

Comment (lauren-li) [reply packages/zowe-explorer-api/src/tree/ZoweVsCodeExtension.ts:32]: @phaumer Got it, thank you for explaining this to me.

File: docs/README-Conformance.md [hunk 16]
```diff
@@ -0,0 +1,312 @@
...
+ <th style=background-color:#5555AA>Criteria </th>
+ </thead>
+
+ <tr>
+   <th style="background-color:#555555">20</th>
+   <th style="background-color:#555555">v1</th>
+   <th style="background-color:#AAAAAA">x</th>
+   <th style="background-color:#AAAAAA"></th>
+   <th></th>
+   <td><b>Menu Names:</b> If the extension is adding new commands and context menu entries to the Zowe Explorer tree view nodes, it adheres to the Zowe Explorer-provided contextual string format.</td>
```
Comment (lauren-li) [inline docs/README-Conformance.md:274]: I thought we had removed this item (20) from the conformance criteria. While it's helpful to extenders that they can hook into Zowe Explorer's contextual string format (and we provide some guidance on this in the Extending readme), I don't think we should restrict them to it. VS Code's API provides quite a few of its own [when clause contexts](https://code.visualstudio.com/api/references/when-clause-contexts) that extenders may also be interested in leveraging.

Comment (crawr) [reply docs/README-Conformance.md]: I have moved this to best practice as discussed during our scrum meeting

**Review by lauren-li — DISMISSED** [review]
I think this PR is in a good state to merge in at this point. I added some minor copyedits to the documentation, and also updated the code samples in `docs/README-Extending.md` to use the latest Zowe Explorer API version available in NPM.

Additional supporting pieces for conformance (e.g. logging) can be added in future PRs, but the conformance table in this PR is complete, and the Extenders' Readme has been updated with a lot of good information that extenders can use right away while other refinements are being made.

Thanks to @phaumer, the Zowe Explorer squad, and everyone else who contributed their perspectives and insights to this PR!

File: docs/README-Extending.md [hunk 17]
```diff
@@ -103,7 +103,7 @@ Currently provided operations:
 - `initForZowe(type: string, meta: ICommandProfileTypeConfiguration[]): Promise<void>;`
 
 ```typescript
-const zoweExplorerApi = ZoweVsCodeExtension.getZoweExplorerApi("1.15.0");
+const zoweExplorerApi = ZoweVsCodeExtension.getZoweExplorerApi("1.16.1");
```
Comment (jellypuno) [inline docs/README-Extending.md:106]: @lauren-li Does this mean that we have to update this const everytime we do a release? or Maybe this is hard-coded for now and it will be changed in the future?

Comment (lauren-li) [reply docs/README-Extending.md:106]: @jellypuno This is just an example. Extenders can specify whichever version of the API they wish to target. I only updated it here because the examples in this document previously had inconsistent version numbers, which could be confusing for readers.

**Review by jellypuno — APPROVED** [review]

**Review by lauren-li — APPROVED** [review]
@jellypuno I have updated the PR to use Zowe Explorer's current view IDs in the Extending README (as PR #1383 will not be merged in for the 1.17.0 release). I have added a footnote about the planned upcoming changes in a to-be-decided future Zowe Explorer release.

All looks good to me now!

**Comment by lauren-li** [comment]
@crawr @jellypuno Per today's scrum discussion, future updates we are considering for `docs/README-Extending.md` include:

- A reference table of the available API features and the earliest version of Zowe Explorer API they became available in.
- For each section describing an API feature, add a line to specify the earliest version of Zowe Explorer API it became available in.
